### ISSUE TITLE
feat: add priority

### ISSUE THREAD

**Issue by tmerse** [MichaelMure/git-bug#72]
Would it be possible prioritize bugs/features (whether alphabetical: "prio:H/M/L" or numerical) and sort accordingly?

**Comment by MichaelMure** [comment]
Interesting. As far as I know, this is a rather uncommon feature of bug trackers. It's also quite subjective.

Could you explain why you need that, give more context ?

A bit unrelated, but have a look at #48

**Comment by neithernut** [comment]
Eh, "priority" and/or "severity" are quite common for "real" bug trackers, e.g. [bugzilla](https://www.bugzilla.org/) or [mantis](http://www.mantisbt.org/). Priority usually expresses the importance of getting a bug fixes/an issue closed, often from the maintainer's point of view. Severity expresses the impact of a bug/issue (e.g. is it only inconvenient, does the application crash or does it corrupt data). Both are often used for sorting issues and deciding what to tackle next.

OT: there are alternatives to maintainer-/user-set priority, e.g. "[user pain](http://www.lostgarden.com/2008/05/improving-bug-triage-with-user-pain.html)".

**Comment by MichaelMure** [comment]
Sorry I wasn't clear.

What I mean is that even if we say the consensus is that it make sense to have this feature (and many bug tracker didn't go that way), it's still something that can be done many ways. Maybe you want 3 levels of priority. Maybe you want 5. Maybe you want to have a specific naming scheme.


There is a reason why there is only the versatile labels, here on github. That doesn't help for sorting though.

To be clear, I'm not saying it shouldn't be done, but I'd like to have a better feeling of what the good way to do it is.

**Comment by muggenhor** [comment]
With most bug trackers that I've worked with they provide some default "priority" and "severity" fields, but they're configurable fields. Maybe supporting such customizable fields instead is good enough?

**Comment by MichaelMure** [comment]
Yes, that is I think the good way to do it: provide a reasonable enough default behavior and have a global scoped configuration if a project want to change the rules.

Have a look at https://github.com/MichaelMure/git-bug/issues/63 for my thoughts on the global config.

**Comment by MichaelMure** [comment]
Note: as a first step, adding a priority feature with a fixed set would be reasonably easy task (extensive but fairly simple). If someone want to tackle that, let me know.

**Comment by muggenhor** [comment]
I expect the fixed set to be a contentious thing though. Unless it can eventually, in a backwards compatible way, be overruled by the config. Making it a number (float or integer) instead and adding a (potentially range-based) mapping to the config (with some sensible default) could help with this.

E.g. priorities <= 1: low, <= 2: medium, <= 3: large.

Just thinking aloud though...

**Comment by MichaelMure** [comment]
> I expect the fixed set to be a contentious thing though.

Ho yes, most likely ...

An important thing: git-bug needs to be able to validate any bug edition at any point in time. This means that:
1. Each bug editions, user profile versions or configurations version will have a Lamport time specifying when they start to be effective
2. User profiles and configurations will be versioned and it will be possible to query, say, what configuration was effective when the priority of a bug was modified.

So, in a way, the question is not really how a set of priority can be backward compatible, but more **how can different set of priorities can coexist for the same bug and across bugs**. As an example, let say that a bug start with a set of priorities, and this set change at some point. As the bug data is immutable, we need to be able to interpret the priorities coherently in the context of a bug, but also across multiple bugs.

**Comment by muggenhor** [comment]
I expect (2) (knowing the config at a specific time) will only help if from the config diff you can automatically produce a function mapping from old to new priorities.

Unless what you mean is to just look at how it appeared historically. I expect comparing bugs with an older priority scheme to those with a new one to be problematic.

**Comment by MichaelMure** [comment]
What I mean is that there is no migration system possible. Once a priority has been set in the history of a bug, it will not change, ever. A new value from a new configured set can override the priority later in the history of the bug though.  This has multiple consequences:

1. in the context of a bug, you can have in its history a priority from an obsolete config set.
1. in the context of a bug, there is not necessarily a value valid in the current config
1. across bugs, you can have one bug with a priority from an outdated config and another with a value from the current config

This means that:

1. we need to be able to validate a config at any point in time (the config system will take care of that)
1. we need a way to compare priorities coming from different and disjoint config
1. we need a way to map the raw value to a user readable thing (number, label, whatever)

I think that your idea of using a number as a the support for priorities would work (it does allow to compare priorities across multiple config). Now the big question is how this number is translated for a user. I suppose we can say that a config is a set of label ("low", "medium", ..), each with a range to interpret the value. But I think we need to also show that a value is from an older config (say "low (outdated)").

This is mainly just rephrasing what you said, but it's kind of important to get the context right. It will also influence how the code works and looks like.

**Comment by MichaelMure** [comment]
But again, we can and should implement this feature with a default fixed set first and make it configurable later.

**Comment by muggenhor** [comment]
> But I think we need to also show that a value is from an older config (say "low (outdated)").

This sounds like a reasonable approach to me. Probably you want to mention exactly which version the config was though.

**Comment by MichaelMure** [comment]
> Probably you want to mention exactly which version the config was though.

Well, that would require to have something meaningful to say about the version though. If we don't add anything specific to the config, we only have the Lamport time (a number) at which it become active, or maybe a hash of the config. That's not really helping. Or we could require a user to name each version of the config, but that would probably leads to garbage name ("config1", "config2" and so on) that wouldn't help either.

... or we could say by how many config version the value is outdated: `low (outdated by 2 versions)`. But that starts to be really verbose. I think I would just say `(outdated)` and wait until it actually becomes a problem to not know by how much.

**Comment by muggenhor** [comment]
I was just thinking a (commit) hash. As long as it's possible to retrieve the config through that it should be enough. The lamport clock doesn't help because it can be equal for different versions produced at the same time.

**Comment by GielVanSchijndel-TomTom** [comment]
*sigh*, stale bots are a terrible idea. This is the perfect example of it. If you want to reject an issue that's fine, just do it. But the *automated* closing via the "stale" argument doesn't fly for feature requests, ever... All it accomplishes is destruction of effort (the thought put into a ticket) and informing people that you're not interested in said effort.

**Comment by sudoforge** [comment]
@GielVanSchijndel-TomTom hey there, thanks for bringing this up.

I initially added the stale configuration as a way to identify things that have been sitting for a long time (there was little to no activity in this repository for two years, after all; and before that, there was about a two year gap between releases as well).

The workflow has [since been refactored](https://github.com/michaelmure/git-bug/pull/1224) such that issues and pull requests will never be closed. This better reflects the intent: that "staleness" is really just an indicator for the maintainers, as a way to say "hey, pay attention to this and take action on it".

**Comment by MichaelMure** [comment]
Not stale

**Comment by abertolli** [comment]
I wanted to add a simple field called "pri" for priority, and it would just take an integer so that people can sort/prioritize however they want.  (e.g. pri 1 would be higher than pri 10)

But I'm not sure how the metadata is structured.  If I want to add a field, where are all the places I need to look at?  I'm not a GO programmer but the syntax seem easy enough.
